internal/utils/ringbuffer/ringbuffer.go

```diff
@@ -0,0 +1,60 @@
+package ringbuffer
+
+type RingBuffer[T any] struct {
+	ring             []T
+	headPos, tailPos int
+	full             bool
+}
+
+func (r *RingBuffer[T]) Init(size int) {
+	r.ring = make([]T, size)
+}
+
+func (r *RingBuffer[T]) Len() int {
+	if r.full {
+		return len(r.ring)
+	}
+	if r.tailPos >= r.headPos {
+		return r.tailPos - r.headPos
+	}
+	return r.tailPos - r.headPos + len(r.ring)
+}
+
+func (r *RingBuffer[T]) PushBack(t T) {
+	if r.full {
+		r.grow()
+	}
+	r.ring[r.tailPos] = t
+	r.tailPos++
+	if r.tailPos == len(r.ring) {
+		r.tailPos = 0
+	}
+	if r.tailPos == r.headPos {
+		r.full = true
+	}
+}
+
+func (r *RingBuffer[T]) PopFront() T {
```

Comment:
What happens if the buffer is empty? Should this be a `Pop() (_ T, ok bool)`?

Reply:
> I was thinking effeciency is all we care about, since it is a internal package.

Kind of, but this is really easy to miss. This only adds a single if statement, which the branch predictor will take care of anyway, so the performance impact is likely negligible.

Do you see a big difference in the benchmark test?